File: src/miner.js
/**
 * 典籍谐音梗挖掘器 (src/miner.js - Node.js JavaScript 版)
 * 职责：读取 data/ 词库与典籍，进行逻辑匹配，导出 dist/xieyin_results.json
 */

const fs = require('fs');
const path = require('path');
const HomophonicEngine = require('./engine');

function loadAllDictionaries(dictDir = path.join(process.cwd(), 'data', 'dictionaries')) {
  const wordsSet = new Set();

  if (fs.existsSync(dictDir)) {
    const files = fs.readdirSync(dictDir);
    for (const fname of files) {
      const fpath = path.join(dictDir, fname);
      if (!fs.statSync(fpath).isFile()) continue;

      const content = fs.readFileSync(fpath, 'utf-8');
      const lines = content.split('\n');

      for (const line of lines) {
        const parts = line.trim().split(/\s+/);
        if (!parts || parts.length === 0) continue;
        const word = parts[0];
        if (/^[\u4e00-\u9fa5]{2,4}$/.test(word)) {
          wordsSet.add(word);
        }
      }
    }
  }

  // 流行文化词汇
  const popCulture = [
    "成龙", "周杰伦", "刘德华", "薛之谦", "沈腾", "贾玲", "周星驰", "张学友", "甄子丹", "徐峥", "坤坤",
    "泰坦尼克", "阿凡达", "流浪地球", "热辣滚烫", "战狼", "满江红", "大话西游", "霸王别姬", "楚门的世界", "泰囧",
    "七里香", "晴天", "稻香", "青花瓷", "双截棍", "卡路里", "小苹果", "孤勇者", "野狼", "告白气球", "奢香夫人",
    "打工人", "程序员", "单身狗", "绝绝子", "降维打击", "双十一", "六一八", "尾款人", "吃瓜群众", "摸鱼侠",
    "涨薪", "支出", "实习", "加仓", "摸鱼", "退款", "下单", "包邮", "离职", "社恐", "社牛", "破防", "加班",
    "吃瓜", "干饭", "白干", "买单", "首付", "有余", "有雨", "通风", "有限", "威武", "低薪", "董事", "砖头"
  ];

  for (const w of popCulture) {
    wordsSet.add(w);
  }

  const finalWords = Array.from(wordsSet);
  console.log(`[Miner JS] 已加载开源词库，共包含 ${finalWords.length} 个常用词汇！`);
  return finalWords;
}

function loadAllCorpus(corpusDir = path.join(process.cwd(), 'data', 'corpus')) {
  const corpusMap = {};
  if (fs.existsSync(corpusDir)) {
    const files = fs.readdirSync(corpusDir).sort();
    for (const fname of files) {
      if (!fname.endsWith('.json')) continue;
      const fpath = path.join(corpusDir, fname);
      const content = fs.readFileSync(fpath, 'utf-8');
      const data = JSON.parse(content);
      const bookName = data.book_name || fname;
      corpusMap[bookName] = data.sentences || [];
    }
  }
  console.log(`[Miner JS] 已加载 ${Object.keys(corpusMap).length} 本古典书籍名篇库！`);
  return corpusMap;
}

function minePuns(outputPath = path.join(process.cwd(), 'dist', 'xieyin_results.json')) {
  const words = loadAllDictionaries();
  const corpus = loadAllCorpus();

  console.log('[Miner JS] 初始化 Node.js 谐音匹配引擎...');
  const engine = new HomophonicEngine(words);

  console.log('[Miner JS] 开始深度挖掘古籍典籍谐音梗...');
  let totalCount = 0;
  const resultsExport = {};

  for (const [bookName, sentences] of Object.entries(corpus)) {
    const bookPuns = [];
    for (const sent of sentences) {
      const puns = engine.findPuns(sent);
      bookPuns.push(...puns);
    }

    // 排序：先同音同调，再按长度降序
    bookPuns.sort((a, b) => {
      if (a.is_same_tone !== b.is_same_tone) {
        return a.is_same_tone ? -1 : 1;
      }
      return b.length - a.length;
    });

    resultsExport[bookName] = bookPuns;
    totalCount += bookPuns.length;
    console.log(`  - ${bookName}: 挖掘到 ${bookPuns.length} 条梗！`);
  }

  const distDir = path.dirname(outputPath);
  if (!fs.existsSync(distDir)) {
    fs.mkdirSync(distDir, { recursive: true });
  }

  fs.writeFileSync(outputPath, JSON.stringify(resultsExport, null, 2), 'utf-8');
  console.log(`[Miner JS] 🎉 成功挖掘出 ${totalCount} 条典籍梗，已导出至 \`${outputPath}\`！`);
  return outputPath;
}

module.exports = minePuns;

if (require.main === module) {
  minePuns();
}

File: src/miner.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
典籍谐音梗挖掘器 (src/miner.py)
职责：纯粹的逻辑计算与梗查找，将分析出来的典籍梗保存导出为 dist/xieyin_results.json
"""

import json
import os
import re
from engine import HomophonicEngine


def load_all_dictionaries(dict_dir="data/dictionaries"):
    words_set = set()

    for fname in os.listdir(dict_dir):
        fpath = os.path.join(dict_dir, fname)
        if not os.path.isfile(fpath):
            continue
        with open(fpath, "r", encoding="utf-8", errors="ignore") as f:
            for line in f:
                parts = line.strip().split()
                if not parts:
                    continue
                word = parts[0]
                if re.match(r"^[\u4e00-\u9fa5]{2,4}$", word):
                    words_set.add(word)

    pop_culture = [
        "成龙",
        "周杰伦",
        "刘德华",
        "薛之谦",
        "沈腾",
        "贾玲",
        "周星驰",
        "张学友",
        "甄子丹",
        "徐峥",
        "坤坤",
        "泰坦尼克",
        "阿凡达",
        "流浪地球",
        "热辣滚烫",
        "战狼",
        "满江红",
        "大话西游",
        "霸王别姬",
        "楚门的世界",
        "泰囧",
        "七里香",
        "晴天",
        "稻香",
        "青花瓷",
        "双截棍",
        "卡路里",
        "小苹果",
        "孤勇者",
        "野狼",
        "告白气球",
        "奢香夫人",
        "打工人",
        "程序员",
        "单身狗",
        "绝绝子",
        "降维打击",
        "双十一",
        "六一八",
        "尾款人",
        "吃瓜群众",
        "摸鱼侠",
        "涨薪",
        "支出",
        "实习",
        "加仓",
        "摸鱼",
        "退款",
        "下单",
        "包邮",
        "离职",
        "社恐",
        "社牛",
        "破防",
        "加班",
        "吃瓜",
        "干饭",
        "白干",
        "买单",
        "首付",
        "有余",
        "有雨",
        "通风",
        "有限",
        "威武",
        "低薪",
        "董事",
        "砖头",
    ]
    for w in pop_culture:
        words_set.add(w)

    final_words = list(words_set)
    print(f"[Miner] 已加载权威开源词库，共包含 {len(final_words)} 个常用词汇！")
    return final_words


def load_all_corpus(corpus_dir="data/corpus"):
    corpus_map = {}
    for fname in sorted(os.listdir(corpus_dir)):
        if not fname.endswith(".json"):
            continue
        fpath = os.path.join(corpus_dir, fname)
        with open(fpath, "r", encoding="utf-8") as f:
            data = json.load(f)
            book_name = data.get("book_name", fname)
            sentences = data.get("sentences", [])
            corpus_map[book_name] = sentences
    print(f"[Miner] 已加载 {len(corpus_map)} 本古典书籍名篇库！")
    return corpus_map


def mine_puns(output_path="dist/xieyin_results.json"):
    words = load_all_dictionaries()
    corpus = load_all_corpus()

    print("[Miner] 初始化谐音匹配引擎...")
    engine = HomophonicEngine(words)

    print("[Miner] 开始在典籍全库中深度挖掘谐音梗...")
    total_count = 0
    results_export = {}

    for book_name, sentences in corpus.items():
        results_export[book_name] = []
        book_puns = []
        for sent in sentences:
            puns = engine.find_puns(sent)
            book_puns.extend(puns)

        sorted_puns = sorted(
            book_puns, key=lambda x: (not x["is_same_tone"], -x["length"])
        )
        results_export[book_name] = sorted_puns
        total_count += len(sorted_puns)
        print(f"  - {book_name}: 挖掘到 {len(sorted_puns)} 条梗！")

    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(results_export, f, ensure_ascii=False, indent=2)

    print(
        f"[Miner] 🎉 成功挖掘出 {total_count} 条典籍梗，已保存至 `{output_path}`！"
    )
    return output_path


if __name__ == "__main__":
    mine_puns()
